### PR TITLE
[#IOCIT-36] Override sender_allowed for Special Services with Activation

### PR DESCRIPTION
**Pull request by BurnedMarshal**
#### List of Changes

`GetLimitedProfile` and `GetLimitedProfileByPost` returns `sender_allowed` property based on `Activation` for special services.

#### Motivation and Context

A special service can use `GetLimitedProfile` to verify the presence of the user into the App IO and the current activation status with only one API call. Nevertheless now the `server_allowed` information is coherent with the status.

#### How Has This Been Tested?

Unit test

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [x] Extend the algorithm for the `PENDING` activation status use case.

### CODE REVIEW

File: utils/__tests__/profile.test.ts [hunk 1]
```diff
@@ -258,14 +285,109 @@ describe("getLimitedProfileTask", () => {
...
+        false,
+        [],
+        mockServicePreferenceModel,
+        mockServiceActivationModel,
+        canSendMessageOnActivation,
+        mockTelemetryClient
+      )();
+      result.apply(mockExpressResponse);
+
+      expect(mockServicePreferenceModel.find).not.toHaveBeenCalled();
```
Comment (michaeldisaro) [inline utils/__tests__/profile.test.ts:346]: I'd check that `mockServiceActivationModel` has been called with something. In normal services tests we checked that it was not called, for special services maybe we can check that it is actually called.

**Review by michaeldisaro — CHANGES_REQUESTED** [review]
Just some minor changes

File: utils/profile.ts [hunk 2]
```diff
@@ -235,52 +242,82 @@ Task<IGetLimitedProfileResponses> =>
...
-          })
+    TE.chain(profile =>
+      pipe(
+        O.fromNullable(userAttributes.service.serviceMetadata),
+        O.map(serviceMetadata => serviceMetadata.category),
+        TE.fromPredicate(
+          category =>
+            O.isSome(category) &&
+            category.value === SpecialServiceCategoryEnum.SPECIAL,
+          identity
```
Comment (fabriziopapi) [inline utils/profile.ts:253]:
```suggestion
               profile,
        TE.fromPredicate(
          () =>
            userAttributes.service.serviceMetadata?.category ===
            SpecialServiceCategoryEnum.SPECIAL,
          identity
```
Could we simplify it with something like this?

Comment (michaeldisaro) [reply utils/profile.ts:253]: I tried it yesterday on another PR too, but I ended thinking that passing from an Option, like @BurnedMarshal did, gives a more precise idea of what we want to obtain. The second one, even if more simple and concise is not declaring an intent imho... it seems more "imperative" to me. But I think it's @BurnedMarshal choice...

File: utils/__tests__/profile.test.ts (hunk 1 above)
Comment (BurnedMarshal) [reply utils/__tests__/profile.test.ts:375]: fixed in [dec75db](https://github.com/pagopa/io-functions-services/pull/215/commits/dec75dbd5fcfb899e0321cd85c761edf3982c68a)

File: utils/__tests__/profile.test.ts [hunk 3]
```diff
@@ -258,14 +285,109 @@ describe("getLimitedProfileTask", () => {
...
+
+      expect(result.kind).toBe("IResponseSuccessJson");
+
+      expect(mockExpressResponse.json).toHaveBeenCalledWith(
+        expect.objectContaining({ sender_allowed: expected })
+      );
+    }
+  );
+
+  it("should responde with an ResponseErrorInternal if an error occours accessing the activation", async () => {
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply utils/__tests__/profile.test.ts]: Fixed in [c16cec7](https://github.com/pagopa/io-functions-services/pull/215/commits/c16cec709c959822eca46a98a08fb154a9daaafa)

**Review by michaeldisaro — APPROVED** [review]

File: utils/services.ts [hunk 4]
```diff
@@ -30,3 +34,29 @@ export const authorizedForSpecialServicesTask = (
...
+ * Configure a function with grace period that
+ * returns true if:
+ * - the Service Activation is present and ACTIVE
+ * - the Service Activation is present and PENDING within the grace period
+ *
+ * @param pendingActivationGracePeriod PENDING grace period in seconds
+ */
+export const canSendMessageOnActivationWithGrace = (
+  pendingActivationGracePeriod: Second
+) => (maybeActivation: O.Option<RetrievedActivation>): boolean =>
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply utils/services.ts]: Fixed in [12fd0ea](https://github.com/pagopa/io-functions-services/pull/215/commits/12fd0ea8d1c8800eecb08f2753cdc173844b5b80)

**Pull request merged**